### DISCUSSION TITLE
How to stub methods in API routes using Cypress?

### DISCUSSION THREAD

**Discussion by michaelmerrill** [vercel/next.js#21394] [Help]
Hi all, I'm having trouble stubbing a method in an api route using NextJS and I was hoping someone could give me some guidance. I've posted this question in cypress discussions as well in case this is a cypress issue.

I want to stub the `validate` method on the Magic admin SDK. It seems like the magic instance I am stubbing in the test is not the same instance that the api handler is using. Any thoughts?

```(typescript)
// lib/magic.ts

import { Magic } from '@magic-sdk/admin';

export const magic = new Magic(process.env.MAGIC_SECRET_KEY);
```

```(typescript)
// cypress/integration/authorization-code.spec.ts
import { magic } from '../../lib/magic';

describe('Authorization Code Flow', () => {
  it('does not redirect user to login page if logged in', function () {
    cy.stub(magic.token, 'validate'); // Magic Admin SDK Error: [ERROR_MALFORMED_TOKEN] The DID token is malformed or failed to parse. (ie. not stubbed)

    cy.request({
      method: 'POST',
      url: '/api/login',
      headers: {
        Authorization: 'Bearer 123',
      },
    });

    cy.visit('/authorize');

    cy.url().should('include', '/authorize');
  });
});
```

```(typescript)
// pages/api/login.ts

import { NextApiRequest, NextApiResponse } from 'next';
import { magic } from '../../lib/magic';
import { setSession } from '../../lib/session';
import { login } from '../../lib/faunadb';

export default async function loginHandler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  try {
    const didToken = magic.utils.parseAuthorizationHeader(
      req.headers.authorization
    );

    magic.token.validate(didToken); // method I'm trying to stub

    const { email } = await magic.users.getMetadataByToken(didToken);

    const user = await login(email);

    await setSession(res, { user });

    res.status(200).send({ done: true });
  } catch (error) {
    res.status(error.status || 500).end(error.message);
  }
}
```

**Comment by tpjnorton** [comment]
I'm running into the same situation - wanting to mock an SDK in an API route but not managing it.

**Comment by Ibtesam-haider96** [comment]
I am also getting the same error while trying to validate DID token on my node.js server. Found any solution?

**Comment by magalidefelippe** [comment]
I have the same problem :(

**Comment by Ibtesam-haider96** [comment]
I was getting this error in my node.js backend. I solved this issue by using passport.js instead of validating DID token in the backend.

**Comment by heymartinadams** [comment]
Running into this issue as well. I just referenced is with Magic. https://github.com/magiclabs/magic-admin-js/issues/86. It appears the issue is with `@magic-sdk/admin` version `1.4.0`.

**Reply by herbievine** [reply]
Just downgraded to `1.3.0` but still not working...

**Reply by heymartinadams** [reply]
Have you tried `1.3.4`? It worked for me at that version.
